Add Try Morello on Linux using Morello IE post

yury-khrustalev opened the pull request.

bsdjhb commented:
In general I like this post and think it covers a fair bit of useful material about CHERI C, mmap, capability permissions, etc.  I have one stylistic suggestion around `PROT_MAX` in your example which would be to maybe tweak it as:

```
    int prot_max = PROT_READ | PROT_WRITE | PROT_EXECUTE;
    int prot = PROT_READ | PROT_WRITE | PROT_MAX(prot_max);
```

Either way is fine though.

My one other general comment is that I would actually avoid talking about hybrid at all.  Historically CHERI began only with hybrid support in the toolchain, but once the toolchain evolved to support purecap C and C++, it is generally much simpler to stick to purecap.  I think for new users it is much simpler to just stick with purecap from the start and only make use of hybrid in special cases.  I'm not sure any of those special cases are common enough for your target audience to warrant talking about hybrid.

jrtc27 commented on _posts/resources/2023-05-08-morello-linux-morelloie.md:
````diff
@@ -0,0 +1,1098 @@
+Note: there is no builtin for LPB and LB object types, you will need to use inline assembly for this:
+
+```
+inline __attribute__ ((naked))
+void *__morello_seal_lpb(void *cap)
+{
+__asm__ volatile(
+"   seal    c0, c0, lpb\n"
+"   ret\n"
+);
+}
````
```suggestion
inline __attribute__ ((naked))
void *__morello_seal_lpb(void *cap)
{
    void *ret;

    __asm__ ("seal %0, %1, lpb" : "=C"(ret) : "C"(cap));
    return (ret);
}
```
is the much less gross way of doing it (not just conceptually; it gives the compiler more flexibility too).

jrtc27 commented on _posts/resources/2023-05-08-morello-linux-morelloie.md:
```diff
@@ -0,0 +1,1098 @@
+}
+
+inline __attribute__ ((naked))
+void *__morello_seal_lb(void *cap)
```
As above

jrtc27 reviewed (commented):
* Please use -march=morello for purecap; +c64 is deprecated
* Please use cheriintrin.h's intrinsics rather than the GNU-specific builtins

yury-khrustalev commented:
Thank you for your comments, I've updated the patch:

- Changed `PROT_MAX` in the example.
- Removed `+c64` from compiler flags.
- Fixed code for `__morello_seal_lpb` and `__morello_seal_lb`.

Regarding the hybrid mode, One of the purposes of this blogpost is to give a wide overview of Morello, so I deliberately added it here since it's an integral part of Morello and there is some interest in it.

Regarding using builtins, again I wanted to show that they exist and at the same time I wanted to keep this small examples as simple and concise as possible while keeping them explicit, hence no usage of `cheriintrin.h`.

rwatson commented on _posts/resources/2023-05-08-morello-linux-morelloie.md:
```diff
@@ -0,0 +1,1096 @@
+
+## Purecap and hybrid
+
+Now it's time to talk about purecap (pure capability) and hybrid Morello applications. From an application developer's point of view, in a purecap program all pointers automatically become capabilities. This means that all memory accesses will happen via (and be controlled by) some specific capability. On the other hand, in a hybrid application, pointers are addresses unless explicitly defined as ...
```
In general, we’ve made a transition to using the terms “CHERI C” and “CHERI C++” when talking about the language-level changes, and “pure cap” (etc) when talking about ABIs, representation of pointers in code, and so on. Hence the CHERI C/C++ programmer’s guide, which leaves open the possibility of different ABIs and code-generation models. I wonder if it could make sense to do the same here?

yury-khrustalev replied:
I will clarify that I use these terms first of all in the context of pointer representation. The angle I want to use here is what happens when we compile (almost) the same code in different ways (making the compiler to invoke different rules and use different ABIs) to use the difference in behaviour to illustrate how capabilities work.

rwatson commented on _posts/resources/2023-05-08-morello-linux-morelloie.md:
```diff
@@ -0,0 +1,1096 @@
+
+Now it's time to talk about purecap (pure capability) and hybrid Morello applications. From an application developer's point of view, in a purecap program all pointers automatically become capabilities. This means that all memory accesses will happen via (and be controlled by) some specific capability. On the other hand, in a hybrid application, pointers are addresses unless explicitly defined as ...
+
+Let's step back and consider these terms: "pointer", "address", and "capability". A pointer is an abstract token that gives you access to some memory. An address is the value of a pointer, i.e. a 64-bit integer that denotes the location in virtual memory. And a capability is one implementation of a pointer. We'll look deeper into capabilities and their properties in the next section. These three  ...
```
This use of the term “value” (although possibly accurate!) might be confusing -- since I guess in pure-capability C/C++, the value of a capability includes its metadata, but the integer cast or interpretation of it, is just the address? But I like the discussion of implementations of a pointers -- I wonder if it’s worth mentioning integer implementations here in the same language, so the reader thinks of them as being different implementations of the same idea?

yury-khrustalev replied:
I will re-word this bit to avoid using the term "value" and instead will use "location [in virtual memory]".

jrtc27 replied:
We have the word address, it means that, please just use it rather than some other clumsy non-standard alternative

yury-khrustalev replied:
I'm using the word address except the place where I explain what "address" is.

rwatson commented on _posts/resources/2023-05-08-morello-linux-morelloie.md:
```diff
@@ -0,0 +1,1096 @@
+
+Let's step back and consider these terms: "pointer", "address", and "capability". A pointer is an abstract token that gives you access to some memory. An address is the value of a pointer, i.e. a 64-bit integer that denotes the location in virtual memory. And a capability is one implementation of a pointer. We'll look deeper into capabilities and their properties in the next section. These three  ...
+
+We're used to so-called "integer pointers" when a pointer doesn't contain any data except the address. From a programmer's point of view, such a pointer can be expressed as `void *`, `size_t`, `uintptr_t` or even `unsigned long`. In Morello, both in purecap and hybrid applications, a pointer and an address are two different things. In purecap, when an unsigned integer type is needed to store a po ...
```
Perhaps “integer implementation of pointers”.

rwatson reviewed (commented):
Not a complete review, but perhaps useful none-the-less?

jrtc27 commented:
> Regarding using builtins, again I wanted to show that they exist and at the same time I wanted to keep this small examples as simple and concise as possible while keeping them explicit, hence no usage of `cheriintrin.h`.

That’s not an excuse to encourage bad practice

yury-khrustalev commented:
> That’s not an excuse to encourage bad practice

@jrtc27, why is this a bad practice, could you please elaborate?

yury-khrustalev commented on _posts/resources/2023-05-08-morello-linux-morelloie.md:
```diff
@@ -0,0 +1,1098 @@
+
+## Setting up your dev environment
+
+Morello comes as a variety of platforms and environments. You can run and experiment with this new CPU architecture on a [Morello prototype board][BOARD], [Morello Platform Model (aka FVP)][FVP], or [Morello Instruction Emulator][MIE].
```
I will mention in for completeness. There are also things like Cheriseed that allow to run and experiment with CHERI code. My limited list was intended to cover all platforms where you can run the same userspace binary without rebuilding it.

yury-khrustalev commented on _posts/resources/2023-05-08-morello-linux-morelloie.md:
```diff
@@ -0,0 +1,1098 @@
+
+Morello platform model provides whole-system emulation in contrast with Morello instruction emulator that only provides userspace emulation. On FVP you can boot an operating system or run baremetal applications, for example, on FVP you can boot and run different Linux distributions. On the other hand, setting up an FVP may be difficult for an inexperienced user, and one may not need full system e ...
+
+Because Morello is backward compatible with AArch64, you don’t need to have the entire distribution built for the Morello target (in fact, many libraries and packages may not be ported yet). What counts is having a kernel that supports Morello at least at a user-kernel ABI level and the userspace can all be AArch64. In an environment like this you would be able to build and run Morello applicatio ...
```
Will re-word it to make clear.

yury-khrustalev commented:
Thanks everyone for the feedback! Hopefully, all is resolved now and this patch can be merged.
